### ISSUE TITLE
limits > 2.0.3 has breaking api changes

### ISSUE THREAD

**Issue by johanjk** [zoltan-fedor/falcon-limiter#1]
The following code works on `limits==2.0.3`:
```python
import falcon_limiter as fl

fl.Limiter(
    key_func=get_access_route_addr,
    default_limits="20 per hour,10 per minute,1 per second",
    config={
        'RATELIMIT_KEY_PREFIX': 'api',
        'RATELIMIT_STORAGE_URL':
            f'redis+sentinel://:{redis.password}@'
            f'{redis.host}:{redis.port}/mymaster',
        'RATELIMIT_STRATEGY': 'moving-window',
        'RATELIMIT_STORAGE_OPTIONS': {
            'sentinel_kwargs': {"password": redis.password}
        },
    }
)
```

When running `limits==2.3.0` I get the following:
```python
  File "/.../python3.9/site-packages/falcon_limiter/middleware.py", line 161, in process_resource
    if not _hit_or_test(_limit, _key):
  File "/.../python3.9/site-packages/limits/strategies.py", line 84, in hit
    return self.storage().acquire_entry(  # type: ignore
  File "/.../python3.9/site-packages/limits/storage/redis.py", line 178, in acquire_entry
    return super()._acquire_entry(key, limit, expiry, self.storage, amount)
  File "/.../python3.9/site-packages/limits/storage/redis.py", line 79, in _acquire_entry
    acquired = self.lua_acquire_window([key], [timestamp, limit, expiry, amount])
  File "/.../python3.9/site-packages/redis/client.py", line 4073, in __call__
    return client.evalsha(self.sha, len(keys), *args)
  File "/.../python3.9/site-packages/redis/client.py", line 3143, in evalsha
    return self.execute_command('EVALSHA', sha, numkeys, *keys_and_args)
  File "/.../python3.9/site-packages/redis/client.py", line 901, in execute_command
    return self.parse_response(conn, command_name, **options)
  File "/.../python3.9/site-packages/redis/client.py", line 915, in parse_response
    response = connection.read_response()
  File "/.../python3.9/site-packages/redis/sentinel.py", line 55, in read_response
    return super(SentinelManagedConnection, self).read_response()
  File "/.../python3.9/site-packages/redis/connection.py", line 739, in read_response
    response = self._parser.read_response()
  File "/.../python3.9/site-packages/redis/connection.py", line 340, in read_response
    raise error
redis.exceptions.AuthenticationError: Authentication required.
```

**Comment by zoltan-fedor** [comment]
Thanks for raising this.

I do not have a Redis Sentinel setup, so I have tested it just with Redis and `limits==2.3.0` works fine just with Redis, so it must be something related to the Sentinel piece.
Let me see if I can find it in the `limits` code.

**Comment by zoltan-fedor** [comment]
I could reproduce the error, but only when the `redis.password` was not provided. See below.

No password:
```
limiter = Limiter(
    key_func=get_access_route_addr,
    default_limits="20 per hour,10 per minute,1 per second",
    config={
        'RATELIMIT_KEY_PREFIX': 'api',
        'RATELIMIT_STORAGE_URL':
            'redis+sentinel://:@127.0.0.1:26379/mymaster',
        'RATELIMIT_STRATEGY': 'moving-window',
        'RATELIMIT_STORAGE_OPTIONS': {
            'sentinel_kwargs': {"password": None}
        },
    }
)
```

Error:
```
2022-01-19 19:06:05 [FALCON] [ERROR] GET /things => Traceback (most recent call last):
  File "falcon/app.py", line 356, in falcon.app.App.__call__
  File "/../falcon_limiter/middleware.py", line 160, in process_resource
    if not _hit_or_test(_limit, _key):
  File "/../python3.8/site-packages/limits/strategies.py", line 84, in hit
    return self.storage().acquire_entry(  # type: ignore
  File "/../python3.8/site-packages/limits/storage/redis.py", line 178, in acquire_entry
    return super()._acquire_entry(key, limit, expiry, self.storage, amount)
  File "/../python3.8/site-packages/limits/storage/redis.py", line 79, in _acquire_entry
    acquired = self.lua_acquire_window([key], [timestamp, limit, expiry, amount])
  File "/../python3.8/site-packages/redis/commands/core.py", line 4440, in __call__
    return client.evalsha(self.sha, len(keys), *args)
  File "/../python3.8/site-packages/redis/commands/core.py", line 3891, in evalsha
    return self.execute_command("EVALSHA", sha, numkeys, *keys_and_args)
  File "/../python3.8/site-packages/redis/client.py", line 1176, in execute_command
    return conn.retry.call_with_retry(
  File "/../python3.8/site-packages/redis/retry.py", line 44, in call_with_retry
    fail(error)
  File "/../python3.8/site-packages/redis/client.py", line 1180, in <lambda>
    lambda error: self._disconnect_raise(conn, error),
  File "/../python3.8/site-packages/redis/client.py", line 1166, in _disconnect_raise
    raise error
  File "/../python3.8/site-packages/redis/retry.py", line 41, in call_with_retry
    return do()
  File "/../python3.8/site-packages/redis/client.py", line 1177, in <lambda>
    lambda: self._send_command_parse_response(
  File "/../python3.8/site-packages/redis/client.py", line 1153, in _send_command_parse_response
    return self.parse_response(conn, command_name, **options)
  File "/../python3.8/site-packages/redis/client.py", line 1192, in parse_response
    response = connection.read_response()
  File "/../python3.8/site-packages/redis/sentinel.py", line 61, in read_response
    return super().read_response(disable_decoding=disable_decoding)
  File "/../python3.8/site-packages/redis/connection.py", line 800, in read_response
    response = self._parser.read_response(disable_decoding=disable_decoding)
  File "/../python3.8/site-packages/redis/connection.py", line 336, in read_response
    raise error
redis.exceptions.AuthenticationError: Authentication required.
```

When the password is provided:
```
limiter = Limiter(
    key_func=get_access_route_addr,
    default_limits="20 per hour,10 per minute,1 per second",
    config={
        'RATELIMIT_KEY_PREFIX': 'api',
        'RATELIMIT_STORAGE_URL':
            'redis+sentinel://:str0ng_passw0rd@127.0.0.1:26379/mymaster',
        'RATELIMIT_STRATEGY': 'moving-window',
        'RATELIMIT_STORAGE_OPTIONS': {
            'sentinel_kwargs': {"password": 'str0ng_passw0rd'}
        },
    }
)
```

Error is different:
```
2022-01-19 19:08:32 [FALCON] [ERROR] GET /things => Traceback (most recent call last):
  File "falcon/app.py", line 356, in falcon.app.App.__call__
  File "/../falcon_limiter/middleware.py", line 160, in process_resource
    if not _hit_or_test(_limit, _key):
  File "/../python3.8/site-packages/limits/strategies.py", line 84, in hit
    return self.storage().acquire_entry(  # type: ignore
  File "/../python3.8/site-packages/limits/storage/redis.py", line 178, in acquire_entry
    return super()._acquire_entry(key, limit, expiry, self.storage, amount)
  File "/../python3.8/site-packages/limits/storage/redis.py", line 79, in _acquire_entry
    acquired = self.lua_acquire_window([key], [timestamp, limit, expiry, amount])
  File "/../python3.8/site-packages/redis/commands/core.py", line 4440, in __call__
    return client.evalsha(self.sha, len(keys), *args)
  File "/../python3.8/site-packages/redis/commands/core.py", line 3891, in evalsha
    return self.execute_command("EVALSHA", sha, numkeys, *keys_and_args)
  File "/../python3.8/site-packages/redis/client.py", line 1173, in execute_command
    conn = self.connection or pool.get_connection(command_name, **options)
  File "/../python3.8/site-packages/redis/connection.py", line 1358, in get_connection
    connection.connect()
  File "/../python3.8/site-packages/redis/sentinel.py", line 54, in connect
    return self.retry.call_with_retry(
  File "/../python3.8/site-packages/redis/retry.py", line 46, in call_with_retry
    raise error
  File "/../python3.8/site-packages/redis/retry.py", line 41, in call_with_retry
    return do()
  File "/../python3.8/site-packages/redis/sentinel.py", line 44, in _connect_retry
    self.connect_to(self.connection_pool.get_master_address())
  File "/../python3.8/site-packages/redis/sentinel.py", line 117, in get_master_address
    master_address = self.sentinel_manager.discover_master(self.service_name)
  File "/../python3.8/site-packages/redis/sentinel.py", line 251, in discover_master
    raise MasterNotFoundError(f"No master found for {service_name!r}")
redis.sentinel.MasterNotFoundError: No master found for 'mymaster'
```

Please verify that the password is in fact provided, as it seems I can only reproduce your error when the password is left empty.

**Comment by johanjk** [comment]
<details>
<summary>Minimal example reproducing the error:</summary>

```python
import falcon
from falcon import testing
import falcon_limiter as fl
from redis import Sentinel
import redis as r
import limits


class redis:
    password = "str0ng_passw0rd"
    host = "redis"
    port = "26379"


# Test that redis is accessible
sentinel = Sentinel(
    sentinels=[(redis.host, redis.port)],
    socket_timeout=0.1,
    sentinel_kwargs={"password": redis.password}
)
try:
    sentinel.discover_master('mymaster')
    sentinel.discover_slaves('mymaster')
except r.sentinel.MasterNotFoundError:
    raise
except r.sentinel.SlaveNotFoundError:
    raise


def get_access_route_addr(req, resp, resource, params) -> str:
    return req.access_route[0]

limiter = fl.Limiter(
    key_func=get_access_route_addr,
    default_limits="20 per hour,10 per minute,1 per second",
    config={
        'RATELIMIT_KEY_PREFIX': 'fotefar-api',
        'RATELIMIT_STORAGE_URL':
            f'redis+sentinel://:{redis.password}@'
            f'{redis.host}:{redis.port}/mymaster',
        'RATELIMIT_STRATEGY': 'moving-window',
        'RATELIMIT_STORAGE_OPTIONS': {
            'sentinel_kwargs': {"password": redis.password}
        },
    }
)

print('falcon', falcon.__version__)
print('falcon_limiter', fl.__version__)
print('limits', limits.__version__)
print('redis', r.__version__)


class Endpoint:
    @limiter.limit()
    def on_post(self, req, resp):
        pass

api = falcon.App(middleware=limiter.middleware)
api.add_route("/foo", Endpoint())

c = testing.TestClient(api)
resp = c.simulate_post('/foo')
```

</details>


<details>
<summary>Output with limits==2.3.0</summary>

```python
The storage backend has failed its check, please verify the provided storage settings!
falcon 3.0.1
falcon_limiter 0.1.2
limits 2.3.0
redis 4.1.1
[FALCON] [ERROR] POST /foo => Traceback (most recent call last):
  File "falcon/app.py", line 356, in falcon.app.App.__call__
  File "/.../python3.9/site-packages/falcon_limiter/middleware.py", line 161, in process_resource
    if not _hit_or_test(_limit, _key):
  File "/.../python3.9/site-packages/limits/strategies.py", line 84, in hit
    return self.storage().acquire_entry(  # type: ignore
  File "/.../python3.9/site-packages/limits/storage/redis.py", line 178, in acquire_entry
    return super()._acquire_entry(key, limit, expiry, self.storage, amount)
  File "/.../python3.9/site-packages/limits/storage/redis.py", line 79, in _acquire_entry
    acquired = self.lua_acquire_window([key], [timestamp, limit, expiry, amount])
  File "/.../python3.9/site-packages/redis/commands/core.py", line 4440, in __call__
    return client.evalsha(self.sha, len(keys), *args)
  File "/.../python3.9/site-packages/redis/commands/core.py", line 3891, in evalsha
    return self.execute_command("EVALSHA", sha, numkeys, *keys_and_args)
  File "/.../python3.9/site-packages/redis/client.py", line 1176, in execute_command
    return conn.retry.call_with_retry(
  File "/.../python3.9/site-packages/redis/retry.py", line 44, in call_with_retry
    fail(error)
  File "/.../python3.9/site-packages/redis/client.py", line 1180, in <lambda>
    lambda error: self._disconnect_raise(conn, error),
  File "/.../python3.9/site-packages/redis/client.py", line 1166, in _disconnect_raise
    raise error
  File "/.../python3.9/site-packages/redis/retry.py", line 41, in call_with_retry
    return do()
  File "/.../python3.9/site-packages/redis/client.py", line 1177, in <lambda>
    lambda: self._send_command_parse_response(
  File "/.../python3.9/site-packages/redis/client.py", line 1153, in _send_command_parse_response
    return self.parse_response(conn, command_name, **options)
  File "/.../python3.9/site-packages/redis/client.py", line 1192, in parse_response
    response = connection.read_response()
  File "/.../python3.9/site-packages/redis/sentinel.py", line 61, in read_response
    return super().read_response(disable_decoding=disable_decoding)
  File "/.../python3.9/site-packages/redis/connection.py", line 800, in read_response
    response = self._parser.read_response(disable_decoding=disable_decoding)
  File "/.../python3.9/site-packages/redis/connection.py", line 336, in read_response
    raise error
redis.exceptions.AuthenticationError: Authentication required.
```

</details>

Output with `limits==2.0.3`:
```
falcon 3.0.1
falcon_limiter 0.1.2
limits 2.0.3
redis 4.1.1
```

With password set to `None` it raises in the initial check:
```python
  File "/.../python3.9/site-packages/redis/sentinel.py", line 251, in discover_master
    raise MasterNotFoundError(f"No master found for {service_name!r}")
redis.sentinel.MasterNotFoundError: No master found for 'mymaster'
```

I am testing on minikube using [redis helm chart](https://github.com/bitnami/charts/tree/master/bitnami/redis) version "15.6.3" with `sentinel.enabled = true`.

**Comment by zoltan-fedor** [comment]
@johanjk, could you please provide the full details for setting up the redis helm chart (by providing the full `values.yam`)?
I assume you set redis password and such.
Thanks

**Comment by zoltan-fedor** [comment]
Thanks @johanjk 

I have managed to install Redis Sentinel on my Kubernetes cluster with your settings and reproduced the issue in `Falcon-Limiter`.

Then I setup a project just using the `limits` library (without `Falcon-Limiter`) and I have managed to reproduce the same there too. So this isn't an issue specific to this library, but it is something with the `limits` library, so I have raised a ticket there, see https://github.com/alisaifee/limits/issues/97

We need to wait for that to be fixed in `limits` or an explanation provided about how you are supposed to supply the password for Redis Sentinel in `2.3.0`.

**Comment by zoltan-fedor** [comment]
HI @johanjk 
The author of the `limits` package identified the issue, see https://github.com/alisaifee/limits/issues/97#issuecomment-1019014760

So this in fact was an unintentional functional change in the `limits` package, which means an undocumented change for all downstream packages too, including  us.

Long story short, now (`limits >= 2.1.0`) for those who are running sentinel with password both on the sentinel and the master/slave, you need to provide both via the `STORAGE_OPTIONS` and the `sentinal_kwargs` instead of the URI, like this:

import falcon_limiter as fl
```

fl.Limiter(
    key_func=get_access_route_addr,
    default_limits="20 per hour,10 per minute,1 per second",
    config={
        'RATELIMIT_KEY_PREFIX': 'api',
        'RATELIMIT_STORAGE_URL':
            f'redis+sentinel://{redis.host}:{redis.port}/mymaster',    # no password in the URI
        'RATELIMIT_STRATEGY': 'moving-window',
        'RATELIMIT_STORAGE_OPTIONS': {
            'sentinel_kwargs': {'password': redis.password},   # sentinel password
            'password': redis.password                         # Redis master/slave password
        },
    }
)
```

Closing this issue. Feel free to reopen if needed.

**Comment by alisaifee** [comment]
Just dropping in to leave a few more trails in case anyone hits this issue. The issue is addressed in [limits 2.3.1](https://limits.readthedocs.io/en/stable/changelog.html#v2-3-1) - which ensures backward compatibility is retained. Additionally, I've updated the documentation a bit to provide more transparency around how the authentication is handled [here](https://limits.readthedocs.io/en/latest/storage.html#redis-with-sentinel)
